src/sentry/relay/config/experimental.py

```diff
@@ -0,0 +1,78 @@
+import logging
+from collections.abc import MutableMapping
+from datetime import datetime, timedelta, timezone
+from typing import Any, Protocol
+
+import sentry_sdk
+
+logger = logging.getLogger(__name__)
+
+
+class TimeoutException(Exception):
+    def __init__(self, elapsed: timedelta, timeout: timedelta, *args: object) -> None:
+        super().__init__(*args)
+        self._elapsed = elapsed
+        self._timeout = timeout
+
+
+class TimeChecker:
+    """Interface to check whether a timeout has been hit.
+
+    The class is initialized with the provided hard timeout. If the timedelta is
+    not bigger than `0`, no checks are performed.  Calling `check` checks the
+    timeout, and raises a `TimeoutException` if it's hit. The timeout starts at
+    the moment the class is initialized.
+    """
+
+    def __init__(self, hard_timeout: timedelta) -> None:
+        self._hard_timeout = hard_timeout
```

Comment:
nit: Why not just call this `timeout`? The "hard" prefix makes me assume there is a soft timeout as well.

Reply:
I intentionally prefixed it to indicate a future action item to add a soft timeout, but happy to rename it if it's confusing.